src/inmanta/deploy/scheduler.py

```diff
@@ -307,7 +333,7 @@ async def _new_version(
                     self._state.drop(resource)
             # Once more, drop all resources that do not exist in this version from the scheduled work, in case they got added
             # again by a deploy trigger (because we dropped them outside the lock).
-            for resource in deleted_resources:
+            for resource in deleted_resources | blocked_resources | transitively_blocked_resources:
                 self._work.delete_resource(resource)
```

Comment:
As implemented now, I don't think blocked resources (especially undefined ones) are safe with respect to races.

The reason we call this method outside of the lock, is that for `deleted_resources`, we don't care too much if we lose the race. The worst that can happen is that it gets deployed once too many. However, if we apply that logic to the blocked resources, its deploy task would choke on the unknowns, wouldn't it? So I think that we have two options:

- make sure that the task can handle the race
- move this call within the lock, for the blocked resources

Reply:
I moved it under the lock.